sky/serve/serve_utils.py

```diff
@@ -0,0 +1,620 @@
+        replica_info = get_replica_info(service_name, with_handle=False)
+        for info in replica_info:
+            if info['replica_id'] == replica_id:
+                return info['status']
+        raise ValueError(
+            _FAILED_TO_FIND_REPLICA_MSG.format(replica_id=replica_id))
+
+    finish_stream = (
+        lambda: _get_replica_status() != serve_state.ReplicaStatus.PROVISIONING)
+    with open(launch_log_file_name, 'r', newline='') as f:
+        for line in _follow_replica_logs(f,
+                                         replica_cluster_name,
+                                         finish_stream=finish_stream,
+                                         exit_if_stream_end=not follow):
+            print(line, end='', flush=True)
+    if (not follow and
+            _get_replica_status() == serve_state.ReplicaStatus.PROVISIONING):
+        # Early exit if not following the logs.
+        return ''
+
+    # Notify user here to make sure user won't think the log is finished.
+    print(f'{colorama.Fore.YELLOW}Start streaming logs for task job '
+          f'of replica {replica_id}...{colorama.Style.RESET_ALL}')
+
+    backend = backends.CloudVmRayBackend()
+    # Always tail the latest logs, which represent user setup & run.
+    returncode = backend.tail_logs(handle, job_id=None, follow=follow)
+    if returncode != 0:
+        return (f'{colorama.Fore.RED}Failed to stream logs for replica '
+                f'{replica_id}.{colorama.Style.RESET_ALL}')
+    return ''
+
+
+def _follow_logs(file: TextIO, *, finish_stream: Callable[[], bool],
+                 exit_if_stream_end: bool) -> Iterator[str]:
+    line = ''
+    while True:
+        tmp = file.readline()
+        if tmp is not None and tmp != '':
+            line += tmp
+            if '\n' in line or '\r' in line:
+                yield line
+                line = ''
+        else:
+            if exit_if_stream_end or finish_stream():
+                break
+            time.sleep(1)
+
+
+def stream_serve_process_logs(service_name: str, stream_controller: bool,
+                              follow: bool) -> str:
+    msg = check_service_status_healthy(service_name)
+    if msg is not None:
+        return msg
+    if stream_controller:
+        log_file = generate_remote_controller_log_file_name(service_name)
+    else:
+        log_file = generate_remote_load_balancer_log_file_name(service_name)
+
+    def _service_is_terminal() -> bool:
+        record = serve_state.get_service_from_name(service_name)
+        if record is None:
+            return True
+        return record['status'] in serve_state.ServiceStatus.failed_statuses()
+
+    with open(os.path.expanduser(log_file), 'r', newline='') as f:
+        for line in _follow_logs(f,
+                                 finish_stream=_service_is_terminal,
+                                 exit_if_stream_end=not follow):
+            print(line, end='', flush=True)
+    return ''
+
+
+class ServeCodeGen:
```

Comment:
Wishing we didn't inherit the codegen pattern here. We used this in CloudVmRayBackend only because we couldn't use Ray Client API for local machine to talk to the remote (it'd restrict the local machine to have the same python/ray versions which is way too restrictive). In SkyServe here, is it easy to replace this with a REST client-server pattern? Worth a TODO.

Reply:
Added a TODO 🫡